onnxruntime/core/providers/nuphar/common/analysis/graph_stats.h

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/codegen/common/common.h"
+#include "core/common/common.h"
+#include "core/graph/graph_viewer.h"
+#include "core/providers/nuphar/common/analysis/analysis.h"
+
+#include "core/providers/nuphar/common/nuphar_subgraph.h"
+// Base class of GraphStatsBase
+// GraphStatsBase holds analysis results from a graph
+// GraphStatsBase can hold multiple analyses
+
+namespace onnxruntime {
+namespace nuphar {
+
+template <typename INPUT_TYPE>
+class GraphStatsBase {
+ public:
+  GraphStatsBase(const std::string& name)
+      : name_(name) {}
+
+  GraphStatsBase() {}
+
+  virtual ~GraphStatsBase() = default;
+
+  // Evaluate all passes
+  virtual void Evaluate(INPUT_TYPE graph) {
+    for (auto& pass : passes_) {
+      pass->Evaluate(graph);
+    }
+  }
+
+  // Set passes externally
+  void SetAllPasses(const std::vector<std::shared_ptr<AnalysisBase<INPUT_TYPE>>>& passes) {
+    passes_.clear();
+    for (auto& pass : passes) {
+      passes_.push_back(pass);
+    }
+  }
+
+  // Set existed evaluated passes externally
+  void SetAllExistedEvaluatedPasses(
+      const std::vector<std::shared_ptr<AnalysisBase<INPUT_TYPE>>>& passes) {
+    existed_eval_passes_.clear();
+    for (auto& pass : passes) {
+      existed_eval_passes_.push_back(pass);
+    }
+  }
+
+  const std::string& Name() const {
+    return name_;
+  }
+
+ protected:
+  // existed eval passes not requiring evaluation
```

Comment:
Can these members be private?

Reply:
Good point, seems not needed for derived class

---
In reply to: [313195838](https://github.com/microsoft/onnxruntime/pull/881#discussion_r313195838) [](ancestors = 313195838)